Worktime tracking for the organization level

kkovacs opened the pull request:
Dear Gitea team,

first of all, thanks for the great work you're doing with this project.

I'm planning to introduce Gitea at a client site, and noticed that while there is time recording, there are no project-manager-friendly reports to actually make use of that data, as were also mentioned by others in #4870 #8684 and #13531.

Since I had a little time last weekend, I had put together something that I hope to be a useful contribution to this great project (while of course useful for me too).

This PR adds a new "Worktime" tab to the Organisation level. There is a date range selector (by default set to the current month), and there are three possible views:

- by repository,
- by milestone, and
- by team member.

Happy to receive any feedback!

There are several possible future improvements of course (predefined date ranges, charts, a member time sheet, matrix of repos/members, etc) but I hope that even in this relatively simple state this would be useful to lots of people.

<img width="1161" alt="Screen Shot 2022-05-25 at 22 12 58" src="https://user-images.githubusercontent.com/118010/170366976-af00c7af-c4f3-4117-86d7-00356d6797a5.png">

Keep up the good work!

Kristof

lafriks reviewed (changes requested):
Thanks for PR! Looks beautiful :)

All queries must be rewritten using xorm query builder and need to add filter to calculate and show only data from repositories user has access to (with issue/pr rights?). We should have function that return such conditions imho

kkovacs commented:
> Thanks for PR! Looks beautiful :)

Thank you, lafriks :) 😊 

> All queries must be rewritten using xorm query builder

OK, if I must :) I'll do it when I'll have some time. Hopefully within a week.

> and need to add filter to calculate and show only data from repositories user has access to (with issue/pr rights?). We should have function that return such conditions imho

Hmm, currently the whole functionality is restricted to organization _"owners"_ (by checking `ctx.Org.IsOwner` in `parseTimes`), and (according to the help text) _"Owners have full access to all repositories and have administrator access to the organization."_ This is why there is no separate rights checking in the query.

My thinking was that I would rather NOT let people see partial (invalid) data, since the risk that such invalid data gets sent out in an invoice etc by somebody who don't have full rights to see everything within an organization is too high.

So I'm double-checking with you: would you consider the current _"limited to owners"_ restriction enough, or still want additional permission check in the query?

Gusted commented:
> So I'm double-checking with you: would you consider the current _"limited to owners"_ restriction enough, or still want additional permission check in the query?

The permission checking shouldn't be done by a helper helper, this should rather be done at router-level. So in this case you want to move the three lines in web.go to the group that starts from line 640, which already includes the owner checking.

lafriks commented:
As @Gusted said it's in the wrong place and that's why I did not notice that it's only for owner team members.

Yes you understand correctly owner users will have full access to all org repositories.

If we leave it that only owners have access to this than it does not need that check in queries.

My use case would be that owner would be either lead developer of the team or devops team. Project manager who would be interested in this tab usually have read rights to all repos (and write rights to issues) thus he would not see this. But that's totally I don't mind seeing done in the future as other PR to improve this feature.

kkovacs commented:
Thanks for the laser-precise advice, @Gusted, will do!

I totally see your point @lafriks. If (as you say) it's OK to deal with the more precise permissions in a future PR, then for simplicity's sake I'd rather leave this particular PR as "owner-only" (but change it of course as per Gusted's recommendation).

I will `push` later!

lafriks commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,208 @@
+		InnerJoin("issue", "tracked_time.issue_id = issue.id").
+		InnerJoin("repository", "issue.repo_id = repository.id").
+		Where(builder.Eq{"repository.owner_id": ctx.Org.Organization.ID}).
+		And(builder.Eq{"tracked_time.deleted": 0}).
```
```suggestion
		And(builder.Eq{"tracked_time.deleted": false}).
```

And same in other places

kkovacs replied:
Thanks, will do!

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,208 @@
+		And(builder.Lt{"tracked_time.created_unix": unixto}).
+		GroupBy("repository.id").
+		OrderBy("repository.name").
+		ToSQL()
```
In general we don't use `ToSQL` and instead just pass this builder into the database engine, which then will handle the juicy bits for us.

kkovacs replied:
Changed!

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,208 @@
+		return
+	}
+	var results []result
+	err = db.GetEngine(ctx).SQL(sql, args...).Find(&results)
```
In this case you want to have something like:
```go
	err = db.GetEngine(ctx).
		Select("repository.name, SUM(tracked_time.time) AS sumtime").
		Table("tracked_time").
		Join("INNER", "issue", "tracked_time.issue_id = issue.id").
		Join("INNER", "repository", "issue.repo_id = repository.id").
		Where(builder.Eq{"repository.owner_id": ctx.Org.Organization.ID}).
		And(builder.Eq{"tracked_time.deleted": false}).
		And(builder.Gte{"tracked_time.created_unix": unixfrom}).
		And(builder.Lt{"tracked_time.created_unix": unixto}).
		GroupBy("repository.id").
		OrderBy("repository.name").
		Find(&results)
```

kkovacs replied:
Thank you @Gusted, I will change it and push!

kkovacs replied:
Changed!

kkovacs commented:
Just two small questions:

1. I'd made the requested changes. Should I request a re-review, or is it enough that they are in the "resolved" state?
2. I've been looking at how you guys implement tests, and I think I could add tests to this functionality if I do some related refactoring for testability. Should I just add the new commits to this PR, or should this be a separate PR later?

Thanks in advance!

lunny commented:
> Just two small questions:
> 
> 1. I'd made the requested changes. Should I request a re-review, or is it enough that they are in the "resolved" state?
> 2. I've been looking at how you guys implement tests, and I think I could add tests to this functionality if I do some related refactoring for testability. Should I just add the new commits to this PR, or should this be a separate PR later?
> 
> Thanks in advance!

Since this PR sent after v1.17 feature freezed, let's wait to merge after v1.17 release. Of course, I think we can re-review it at any time when maintainers have free time.

kkovacs commented:
Thanks, I'll be requesting re-review then. (Github is a bit confusing for me regarding the right etiquette and the request UI)

kkovacs commented:
Dear @lafriks and @Gusted, I just couldn't bear not having tests on the correctness of the queries, so I refactored a bit and added tests. Hope you don't mind.

P.s.: The build failed with `"https://registry-1.docker.io/v2/woodpeckerci/plugin-codecov/manifests/next-alpine": received unexpected HTTP status: 500 Internal Server Error`, which sounds like a docker registry issue? Can the build be restarted?

P.s.2: I think Github erroneously keeps "1 change requested" on this ticket, because the change *was* done – maybe it was confused by the _rebase_ at the same time?

Anyway, I will leave this PR alone for now -- feel free to review (all change requests were completed) and to merge at the right time!

Gusted commented on modules/util/sec_to_hour.go:
```diff
@@ -0,0 +1,39 @@
+
+	// Extract only the relevant information of the time
+	switch {
+	case hours > 0:
```
This should be a simple `if { ... } else { ... }`

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+
+// resultTimesByRepos is a struct for DB query results
+type resultTimesByRepos struct {
+	Name    string `xorm:"name"`
+	SumTime int64  `xorm:"sumtime"`
+}
+
+// resultTimesByMilestones is a struct for DB query results
+type resultTimesByMilestones struct {
+	RepoName     string `xorm:"reponame"`
+	Name         string `xorm:"name"`
+	ID           string `xorm:"id"`
+	SumTime      int64  `xorm:"sumtime"`
+	HideRepoName bool   `xorm:"-"`
+}
+
+// resultTimesByMembers is a struct for DB query results
+type resultTimesByMembers struct {
+	Name    string `xorm:"name"`
+	SumTime int64  `xorm:"sumtime"`
```
Usually you only use `xorm` tags for struct that will be stored into the database. Xorm should be able to figure out by itself which row corresponds to a field(if not, then add a xorm tag).

kkovacs replied:
Changed. To let Xorm figure out by itself, the query builder part was also touched (`sumtime` => `sum_time`, so it recognizes `SumTime`, etc.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+	ctx.Data["RangeTo"] = to
+
+	// Prepare unix time values for SQL
+	gfrom, err := time.Parse("2006-01-02", from)
```
what does the `g` represent in the variable?

kkovacs replied:
It stands for "Go", as in Go's Time object. The variable named `from` above is of String type, and here we convert to Go's Time. It's a temporary variable that is needed for the handling of `err` from the `Parse` function. Changed it to `from2`.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+	unixfrom = gfrom.Unix()
+	gto, err := time.Parse("2006-01-02", to)
+	if err != nil {
+		ctx.ServerError("TimesError", err)
```
```suggestion
		ctx.ServerError("time.Parse", err)
		return
```

kkovacs replied:
Changed.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+	// Prepare unix time values for SQL
+	gfrom, err := time.Parse("2006-01-02", from)
+	if err != nil {
+		ctx.ServerError("TimesError", err)
```
```suggestion
		ctx.ServerError("time.Parse", err)
		return
```

kkovacs replied:
Changed.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+
+	results, err := getTimesByRepos(unixfrom, unixto, ctx.Org.Organization.ID)
+	if err != nil {
+		ctx.ServerError("TimesError", err)
```
```suggestion
		ctx.ServerError("getTimesByRepos", err)
```

kkovacs replied:
Changed.

Gusted commented on routers/web/web.go:
```diff
@@ -642,6 +642,10 @@ func RegisterRoutes(m *web.Route) {
 			m.Post("/teams/{team}/edit", bindIgnErr(forms.CreateTeamForm{}), org.EditTeamPost)
 			m.Post("/teams/{team}/delete", org.DeleteTeam)
 
+			m.Get("/times/by_repos", org.TimesByRepos)
+			m.Get("/times/by_members", org.TimesByMembers)
+			m.Get("/times/by_milestones", org.TimesByMilestones)
```
Please use `m.Group` here.

kkovacs replied:
Changed. Please review for correct use of `context.OrgAssignment`!

Gusted commented on templates/org/times/daterange.tmpl:
```diff
@@ -0,0 +1,15 @@
+<div class="ui form">
+	<div class="field">
+		<label>{{$.i18n.Tr "Start date"}}</label>
```
Please use the key specified in the `options/locale/locale_en-US.ini` file.

kkovacs replied:
Changed.

Gusted commented on templates/org/times/daterange.tmpl:
```diff
@@ -0,0 +1,15 @@
+		</div>
+	</div>
+	<div class="field">
+		<label>{{$.i18n.Tr "End date"}}</label>
```
Ditto.

kkovacs replied:
Changed.

Gusted commented on templates/org/times/daterange.tmpl:
```diff
@@ -0,0 +1,15 @@
+			<input type="date" id="rangeto" name="to" value="{{.RangeTo}}">
+		</div>
+	</div>
+	<button class="ui primary button">{{$.i18n.Tr "Update"}}</button>
```
Ditto

Gusted commented on templates/org/times/submenu.tmpl:
```diff
@@ -0,0 +1,8 @@
+<div class="ui column">
+	<div class="ui compact small menu">
+
```
```suggestion
```

kkovacs replied:
Changed.

Gusted commented on templates/org/times/submenu.tmpl:
```diff
@@ -0,0 +1,8 @@
+<div class="ui column">
+	<div class="ui compact small menu">
+
+		<a class="{{if .TabIsByRepos}}active{{end}} item" href="{{$.Org.OrganisationLink}}/times/by_repos?from={{.RangeFrom}}&to={{.RangeTo}}">{{svg "octicon-repo" 16 "mr-2"}} {{$.i18n.Tr "By repositories"}}</a>
+		<a class="{{if .TabIsByMilestones}}active{{end}} item" href="{{$.Org.OrganisationLink}}/times/by_milestones?from={{.RangeFrom}}&to={{.RangeTo}}">{{svg "octicon-milestone" 16 "mr-2"}} {{$.i18n.Tr "By milestones"}}</a>
+		<a class="{{if .TabIsByMembers}}active{{end}} item" href="{{$.Org.OrganisationLink}}/times/by_members?from={{.RangeFrom}}&to={{.RangeTo}}">{{svg "octicon-people" 16 "mr-2"}} {{$.i18n.Tr "By members"}}</a>
```
Please use correct translation key here as well.

Gusted reviewed (commented):
Please make sure you have translated words and use the correct keys for those. The HTML part of this PR really need some tidying up to do.

kkovacs commented:
Thanks for the detailed review - I haven't disappeared, I'll do the requested changes when I have some time!

kkovacs commented:
Dear @Gusted, I've finally got around to make the changes you requested.

It's not completely clear to me whether it should be me or you who "resolve" the conversations, so I erred on the side of "resolving" the trivial ones, and leaving open the ones where you might want to see the responses.

One particular thing: please review my use of `context.OrgAssignment` for the `m.Group` in routers.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+
+// parseTimes contains functionality that is required in all these functions,
+// like parsing the date from the request, setting default dates, etc.
+func parseTimes(ctx *context.Context) (unixfrom, unixto int64, err error) {
```
`parseOrgTimes`.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+func parseTimes(ctx *context.Context) (unixfrom, unixto int64, err error) {
+	// View variables
+	ctx.Data["PageIsOrgTimes"] = true
+	ctx.Data["AppSubURL"] = setting.AppSubURL
```
This shouldn't be set by this function.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+		OrderBy("repository.name, milestone.deadline_unix, milestone.id").
+		Find(&results)
+
+	// Show only the first RepoName, for nicer output.
+	prevreponame := ""
+	for i := 0; i < len(results); i++ {
+		res := &results[i]
+		if prevreponame == res.RepoName {
+			res.HideRepoName = true
+		}
+		prevreponame = res.RepoName
+	}
```
This should be moved outside this function.

Gusted commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,218 @@
+}
+
+// getTimesByRepos fetches data from DB to serve TimesByRepos.
+func getTimesByRepos(unixfrom, unixto, orgid int64) (results []resultTimesByRepos, err error) {
```
These kind of functions lives in `models/` directory. `routers/` is just connecting all the functions and do the nice things with the data.

Gusted commented on routers/web/web.go:
```diff
@@ -642,6 +642,12 @@ func RegisterRoutes(m *web.Route) {
+				m.Get("/by_repos", org.TimesByRepos)
+				m.Get("/by_members", org.TimesByMembers)
+				m.Get("/by_milestones", org.TimesByMilestones)
+			}, context.OrgAssignment(true, true))
```
```suggestion
			}, context.OrgAssignment(false, true))
```
Only org admin should be able to see it right?

Gusted commented on templates/org/menu.tmpl:
```diff
@@ -21,6 +21,11 @@
 			</a>
+			{{if .IsOrganizationOwner}}
+				<a class="{{if $.PageIsOrgTimes}}active{{end}} item" href="{{$.OrgLink}}/times/by_repos">
+					{{svg "octicon-clock"}}&nbsp;{{$.i18n.Tr "org.worktime"}}
```
Please change all occurences of `$.i18n.Tr` to `$.locale.Tr`

Gusted commented:
Tests are failing as you need to make an authenticated request with a user

<details>

```
--- FAIL: TestOrgMembers (1.41s)

    org_test.go:124: 

        	Error Trace:	integration_test.go:506

        	            				org_test.go:124

        	Error:      	Not equal: 

        	            	expected: 200

        	            	actual  : 303

        	Test:       	TestOrgMembers

        	Messages:   	Request: GET /org/org25/members

    testlogger.go:78: 2022/08/11 14:58:27 ...eb/routing/logger.go:99:func1() [I] [62f51913] router: completed GET /org/org25/members for , 303 See Other in 0.4ms @ context/auth.go:28(context.Toggle)

    org_test.go:124: Response: <a href="/user/login">See Other</a>.

        

        

    testlogger.go:78: 2022/08/11 14:58:27 ...eb/routing/logger.go:99:func1() [I] [62f51913-2] router: completed GET /user/login for , 200 OK in 7.8ms @ auth/auth.go:150(auth.SignIn)

    testlogger.go:78: 2022/08/11 14:58:27 ...eb/routing/logger.go:99:func1() [I] [62f51913-3] router: completed POST /user/login for , 303 See Other in 342.8ms @ auth/auth.go:175(auth.SignInPost)

    org_test.go:129: 

        	Error Trace:	integration_test.go:506

        	            				integration_test.go:335

        	            				org_test.go:129

        	Error:      	Not equal: 

        	            	expected: 200

        	            	actual  : 303

        	Test:       	TestOrgMembers

        	Messages:   	Request: GET /org/org25/members

    org_test.go:129: Response: <a href="/user/login">See Other</a>.

        

        

    testlogger.go:78: 2022/08/11 14:58:27 ...eb/routing/logger.go:99:func1() [I] [62f51913-4] router: completed GET /org/org25/members for , 303 See Other in 0.8ms @ context/auth.go:28(context.Toggle)

    testlogger.go:78: 2022/08/11 14:58:27 ...nization/org_user.go:108:loadOrganizationOwners() [E] [62f51913-5] Organization does not have owner team: 25

    testlogger.go:78: 2022/08/11 14:58:27 ...eb/routing/logger.go:99:func1() [I] [62f51913-5] router: completed GET /org/org25/members for , 200 OK in 27.8ms @ org/members.go:25(org.Members)
```

</details>

kkovacs commented:
Sorry it took this long to fix this test-fail, it was a bit hard to reproduce on my machine. I learned a lot about drone :) I rebased and hopefully now every test will succeed.

lunny commented on models/organization/org_time/org_times_test.go:
```diff
@@ -0,0 +1,305 @@
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package orgtime
```
Use `orgtime_test` to avoid dependencies cycle.

kkovacs replied:
Fixed!

lunny commented on models/organization/org_times.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+// GetTimesByRepos fetches data from DB to serve TimesByRepos.
+func (org *Organization) GetTimesByRepos(unixfrom, unixto int64) (results []ResultTimesByRepos, err error) {
```
They don't need to become methods of Organization and they should get `context.Context` as the first parameter.

kkovacs replied:
Hmm, if the functions would not be a method of Organization, then it would have to take an `org_id` as a parameter, since it calculates on an Origanization basis (not globally). Would that be preferred, or should it stay an object function?

lunny replied:
The Organization object could be the first parameter of the function.  Thinking the difference between `func (org *Organization) GetTimesByRepos` and `func GetTimesByRepos(org *Organization`, the previous should be used when they are almost handling a property, otherwise use later. Of course, many places still didn't follow that which we need to do refactor.

molaie commented:
@kkovacs thanks for this great work.
keep your excellent work.

lunny commented on models/organization/org_time/org_times_test.go:
```diff
@@ -0,0 +1,305 @@
+func TestMain(m *testing.M) {
+	unittest.MainTest(m, &unittest.TestOptions{
+		GiteaRootPath: filepath.Join("..", "..", ".."),
+		FixtureFiles: []string{
```
The FixtureFiles here is unnecessary any more

kkovacs replied:
I don't know, if I take that `FixtureFiles` out, the tests fail with:

```
=== RUN   TestTimesPrepareDB
LoadFixtures failed after retries: testfixtures: could not clean table "action": no such table: action
    org_times_test.go:27:
                Error Trace:    /persist/gitea/src/models/organization/org_time/org_times_test.go:27
                Error:          Received unexpected error:
                                testfixtures: could not clean table "action": no such table: action
                Test:           TestTimesPrepareDB
--- FAIL: TestTimesPrepareDB (4.02s)
=== RUN   TestTimesByRepos
=== RUN   TestTimesByRepos/Full_sum_for_org_1
    org_times_test.go:104:
                Error Trace:    /persist/gitea/src/models/organization/org_time/org_times_test.go:104
                Error:          Received unexpected error:
                                user does not exist [uid: 1, name: , keyid: 0]
                Test:           TestTimesByRepos/Full_sum_for_org_1
--- FAIL: TestTimesByRepos (0.00s)
    --- FAIL: TestTimesByRepos/Full_sum_for_org_1 (0.00s)
```

Please advise!

lunny replied:
Since now almost all tests has a different package name with non-test code in the same directory. We can import the package where action is in.

kkovacs replied:
Thanks for the feedback, but sorry, I'm not sure I understand it. Yes, we can import the package ( `code.gitea.io/gitea/models/organization`, where the tested functions are, is already imported), but how does that help against the test data (fixtures) tables not existing for the tests, if I remove the FixtureFiles like you asked?

lunny replied:
Like you import `_ "code.gitea.io/gitea/models/issues"`, The init() function will `RegisterModel` the necessary model for this test package.

lunny commented on models/organization/org_times.go:
```diff
@@ -0,0 +1,86 @@
+
+import (
+	"code.gitea.io/gitea/models/db"
+	"xorm.io/builder"
```
blank line.

kkovacs replied:
I'm sorry I don't understand :) Does it *need* a blank line? (Where?) Does it *have* an unnecessary blank line?

(`lint` and `gofrumpt` does not complain.)

jolheiser replied:
Generally as a convention in this code base we group imports in three blocks

```go
import (
    // stdlib

    // packages in this repo

    // packages from other modules
)
```

So a blank line between these imports would be our convention. 🙂

kkovacs replied:
Ah, thank you! I was not aware of the _second_ blank line.

lunny commented on routers/web/org/times.go:
```diff
@@ -0,0 +1,147 @@
+// Copyright 2014 The Gogs Authors. All rights reserved.
```
This line is unnecessary.

bard commented:
This looks mighty useful! Other than rebasing, are the changes required at this point only style-related?

kkovacs commented:
> This looks mighty useful! Other than rebasing, are the changes required at this point only style-related?

Thank you for the kind words, @bard! :) Nice comments like yours keep open source projects alive.

But, unfortunately, no. What is asked is another (the 5th) refactoring of the code. Also, the `main` branch in the meantime has diverged enough that rebasing the code is now at least half-day task in itself. It would be easier to integrate the whole stuff again. (I did try a quick rebase attempt before writing this, to see how much work was needed.)

Now, it's normal to have 1-2 rounds of change requests on a PR -- it is very important to keep the structure of a project consistent, enforce style so it's easy for everyone to touch each part, etc.

The problem is that when there were already 5 rounds, AND each round is asking for NEW things to change which were NOT mentioned before, I see no point in doing the integration work again, since there is no guarantee that it will not be in vain again. I unfortunately don't have THIS much free time on my hand. :( I've already spent about four times as much time complying with review requests as on the original feature itself...

The code still stands contributed, if someone is willing to integrate it in a way that satisfies the code stewards, that's great, I would be the most happy -- I too, still believe that this would increase the value of `gitea` a lot. I also had other project management centric ideas to expand on this feature that I planned to contribute in later separate PRs, but I wanted to keep it simple first. It didn't really work out :)

Sorry.

lunny commented:
> > This looks mighty useful! Other than rebasing, are the changes required at this point only style-related?
> 
> Thank you for the kind words, @bard! :) Nice comments like yours keep open source projects alive.
> 
> But, unfortunately, no. What is asked is another (the 5th) refactoring of the code. Also, the `main` branch in the meantime has diverged enough that rebasing the code is now at least half-day task in itself. It would be easier to integrate the whole stuff again. (I did try a quick rebase attempt before writing this, to see how much work was needed.)
> 
> Now, it's normal to have 1-2 rounds of change requests on a PR -- it is very important to keep the structure of a project consistent, enforce style so it's easy for everyone to touch each part, etc.
> 
> The problem is that when there were already 5 rounds, AND each round is asking for NEW things to change which were NOT mentioned before, I see no point in doing the integration work again, since there is no guarantee that it will not be in vain again. I unfortunately don't have THIS much free time on my hand. :( I've already spent about four times as much time complying with review requests as on the original feature itself...
> 
> The code still stands contributed, if someone is willing to integrate it in a way that satisfies the code stewards, that's great, I would be the most happy -- I too, still believe that this would increase the value of `gitea` a lot. I also had other project management centric ideas to expand on this feature that I planned to contribute in later separate PRs, but I wanted to keep it simple first. It didn't really work out :)
> 
> Sorry.

Thanks for your time! What a pity this cannot be merged ASAP. I wish someone could pick it and continue.

kkovacs commented:
OK, I'm giving this another try, because you guys were so nice. I've rebased it, implemented the last change request, and the build is passing.

If there are any more change requests, say so, but please try to ask in one batch, not with a thousand cuts, please? Thanks :)

bauermarkus commented:
> OK, I'm giving this another try, because you guys were so nice. I've rebased it, implemented the last change request, and the build is passing.
> 
> If there are any more change requests, say so, but please try to ask in one batch, not with a thousand cuts, please? Thanks :)

Thank you a lot for your work and your effort! That's such a wonderful feature!

bard commented:
@lunny @delvh (paging because I see you in the PR, sorry if you're not the ones I should @): looking at this and #23113, both in limbo, it seems that time tracking in general in Gitea is on hold. Obviously I'm among those here who believe that small additions in this area would yield disproportionately high end user value — teams wouldn't have to find, integrate, and maintain a separate time tracking solution — but I respect that no software can be everything to everyone.

Any chance we could have the maintaners' view on the time tracking story in Gitea? At least we can let the wish go for good if necessary.

delvh commented:
The problems I have with time tracking are the following:
1. I can see the benefit of allowing it for Gitea. However, if it is introduced without a clear roadmap of how it interacts with everything, it will simply be a burden. As @jolheiser would put it, before the whole functionality and its roadmap go through an RFC process (which we haven't established yet, so it would mostly be a long issue discussion at the moment), I don't think it's a good idea to simply merge it. It will create too much technical debt.
2. It is quite a struggle between keeping Gitea un-bloated and adding new features. While we mostly answer it with `yes` for any git/issue-related question, project-management features are another thing. While I too would like a better project management integration as GitHub does, I also know that it will drastically change the focus of Gitea, and thus be a lot of work.
3. As you can see, we currently have 2000 issues and 170 PRs open. Many good features are already ignored, simply because we cannot keep up with how much work there is to do. Or as I call it: The void of ignored PRs starting on page 2. The result of that is in praxis that larger PRs tend to be forgotten, especially if they introduce a not well-thought-out feature. No one wants to review that, there are more interesting/new/urgent PRs to review.

bard commented:
That's fair. You see value in this (ancillary, I agree) area, but there are more urgent issues from the core area, and the lack of RFC process makes it difficult to achieve a well-integrated change, and the size of proposed changes makes it difficult to assess for impact and review.

I guess that for things to progress in the short term it would take either someone who has both a good understanding of the code base and a carefully thought-out story for project management that maintainers agree with (unlikely, as that sounds more like a core contributor and core contributors already have their hands full), or changes that are tiny and incremental enough that are easy to review and assess as well-fitting (I don't know if features in this area can actually be shaped this way).

Not great news obviously it helps a lot with setting expectations, so thank you for chiming in.

wxiaoguang commented:
> Thanks for PR! Looks beautiful :)
> 
> All queries must be rewritten using xorm query builder and need to add filter to calculate and show only data from repositories user has access to (with issue/pr rights?). We should have function that return such conditions imho

@lafriks The queries are in XORM now, so your "change request" could be dismissed?

wxiaoguang commented:
Made some changes, and the UI is kept as before

<details>

![image](https://github.com/user-attachments/assets/d61d39b0-ec66-47ef-8819-d789a97e4f0d)

----

![image](https://github.com/user-attachments/assets/a16a6bd9-5810-4337-b6e8-b9bc3e71d5b0)

----

![image](https://github.com/user-attachments/assets/8c30619a-f068-4a45-840c-133026c0a664)

</details>

wxiaoguang commented:
After fixing some problems (mismatched HTML tags, different SQL syntaxes), I think it is good enough now.

CI passes.
